src/ServiceControl/Recoverability/Retrying/RetryProcessor.cs

```diff
@@ -195,6 +198,57 @@ async Task<int> Stage(RetryBatch stagingBatch, IAsyncDocumentSession session)
             return messages.Length;
         }
 
+        async Task TryStageMessage(FailedMessage message, string stagingId, FailedMessageRetry failedMessageRetry)
+        {
+            try
+            {
+                await StageMessage(message, stagingId).ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                if (failedMessageRetry.StageAttempts < 5)
+                {
+                    await IncrementAttemptCounter(failedMessageRetry).ConfigureAwait(false);
+                    //We need to rethrow here to stop the current staging process
+                    throw new Exception($"Error while trying to stage failed message {message.UniqueMessageId}", e);
+                }
+
+                Log.Error($"Failed message {message.UniqueMessageId} reached its staging retry limit ({failedMessageRetry.StageAttempts}) and is going to be removed from the batch.", e);
```

Comment:
Not sure if I should use `Id` or `UniqueMessageId` when logging 😕

Reply:
UniqueMessageId is the document Id isn't it? It will make it easier to find in Maintenance mode. Also, Message Id is non-unique. You could have an event (one Id) handled by multiple subscribers. Each one gets it's own UniqueMessageId. You can think of it as a Logical Message Processing Id.